server/repository/src/db_diesel/temperature_breach.rs

```diff
@@ -87,37 +98,68 @@ impl<'a> TemperatureBreachRepository<'a> {
             .limit(pagination.limit as i64)
-            .load::<TemperatureBreachRow>(&self.connection.connection)?;
+            .load::<TemperatureBreachJoin>(&self.connection.connection)?;
 
         Ok(result.into_iter().map(to_domain).collect())
     }
-}
 
-type BoxedLogQuery = temperature_breach::BoxedQuery<'static, DBType>;
-
-fn create_filtered_query(filter: Option<TemperatureBreachFilter>) -> BoxedLogQuery {
-    let mut query = temperature_breach::table.into_boxed();
-
-    if let Some(filter) = filter {
-        apply_equal_filter!(query, filter.id, temperature_breach_dsl::id);
-        apply_equal_filter!(query, filter.sensor_id, temperature_breach_dsl::sensor_id);
-        apply_equal_filter!(query, filter.store_id, temperature_breach_dsl::store_id);
-        //apply_equal_filter!(query, filter.r#type, temperature_breach_dsl::type_);
-        apply_date_time_filter!(
-            query,
-            filter.start_timestamp,
-            temperature_breach_dsl::start_timestamp
-        );
-        apply_date_time_filter!(
-            query,
-            filter.end_timestamp,
-            temperature_breach_dsl::end_timestamp
-        );
-    }
-
-    query
+    pub fn create_filtered_query(
+        filter: Option<TemperatureBreachFilter>,
+    ) -> Result<BoxedTemperatureBreachQuery, RepositoryError> {
+        let mut query = temperature_breach_dsl::temperature_breach
+            .inner_join(sensor_dsl::sensor)
+            .left_join(location_dsl::location)
+            .into_boxed();
+
+        if let Some(f) = filter {
+            let TemperatureBreachFilter {
+                id,
+                store_id,
+                acknowledged,
+                start_timestamp,
+                end_timestamp,
+                r#type,
+                sensor,
+                location,
+            } = f;
+
+            apply_equal_filter!(query, id, temperature_breach_dsl::id);
+            apply_equal_filter!(query, store_id, temperature_breach_dsl::store_id);
+            apply_equal_filter!(query, r#type, temperature_breach_dsl::type_);
+            apply_date_time_filter!(
+                query,
+                start_timestamp,
+                temperature_breach_dsl::start_timestamp
+            );
+            apply_date_time_filter!(query, end_timestamp, temperature_breach_dsl::end_timestamp);
+
+            if let Some(value) = acknowledged {
+                query = query.filter(temperature_breach_dsl::acknowledged.eq(value));
+            }
+
+            if sensor.is_some() {
+                let sensor_ids =
+                    SensorRepository::create_filtered_query(sensor).select(sensor_dsl::id);
+                query = query.filter(temperature_breach_dsl::sensor_id.eq_any(sensor_ids));
+            }
+
+            if location.is_some() {
+                let location_ids = LocationRepository::create_filtered_query(location)
+                    .select(location_dsl::id.nullable());
```

Comment:
Why do you use `nullable()` here? what was the problem? I see the error message but it isn't helpful for me. How do found out about it?

Reply:
It's because location_id is a nullable field and null can't be iterated over, so had to convert it into a non-null value